### PR TITLE
build: update to walletdb/bdb version w/ mmap pre-sizing

### PR DESCRIPTION
**Pull request by Roasbeef**
In this commit, we update to the `walletdb/bdb` version that now has
mmap-pre-sizing. By sizing the memory map ahead of time, we aim to avoid
the mass copies in the DB each time we need to expand the size of the
DB. The new version will set the mmap size to 2x the db size, which
allows the DB to double before we need to remap everything.

### CODE REVIEW

**Comment by joostjager** [comment]
Is there a PR on btcwallet to review, or do you request me to just review the lnd changes here?

File: channeldb/db.go [hunk 1]
```diff
@@ -165,7 +166,10 @@ func Open(dbPath string, modifiers ...OptionModifier) (*DB, error) {
 
 	// Specify bbolt freelist options to reduce heap pressure in case the
 	// freelist grows to be very large.
-	bdb, err := kvdb.Open(kvdb.BoltBackendName, path, opts.NoFreelistSync)
+	bboltOpts := &bbolt.Options{
```
Comment (bhandras) [inline channeldb/db.go:169]: Q: In the etcd PR `NoFreeListSync` is an `lncfg` option. I wonder if we ever intend to actually set this to false? And if no, then perhaps we could default it somewhere (walletdb?) such that it's less hurdle to later on remove them from the code.

File: channeldb/db.go [hunk 2]
```diff
@@ -251,7 +255,10 @@ func createChannelDB(dbPath string) error {
 	}
 
 	path := filepath.Join(dbPath, dbName)
-	bdb, err := kvdb.Create(kvdb.BoltBackendName, path, true)
+	bboltOpts := &bbolt.Options{
+		NoFreelistSync: true,
```
Comment (Crypt-iQ) [inline channeldb/db.go:259]: Though this is just for creating the db, should this be user-specified as well? Setting it one way or another shouldn't give any performance benefit as:
- if we sync the freelist (this takes time), the next startup upon `Open` will be shorter
- if we don't sync the freelist (quick `Create` time), the next startup upon `Open` will be longer as it needs to construct the freelist

So both scenarios are quicker in one area and slower in another. Only reason I think would be for consistency.

File: channeldb/db.go (hunk 1 above)
Comment (Crypt-iQ) [reply channeldb/db.go:169]: I think we want this to be user-configurable as some users have reported slow startup times on weak machines (mobile, rpi) and opt to not set `NoFreeListSync`.

**Comment by Roasbeef** [comment]
@joostjager yeah that's here: https://github.com/btcsuite/btcwallet/pull/697

**Comment by Roasbeef** [comment]
Pushed up two small optimizations to the migration:
  1. Re-use the temp variables that remain the same and or are used just to access keys. 
 2. Remove the usage of `append` for `attemptInfo` as this is where most of the memory is being allocated. As is, [it'll end up doubling the size of the slice each time to make room for future appends](https://github.com/golang/go/blob/master/src/runtime/slice.go#L96). In our case, there will be no future appends, so we can just copy the exact amount of bytes we need. 


For reference, here's a heap profile of the migration as is: 
```
(pprof) list MigrateMPP
Total: 1.52GB
ROUTINE ======================== github.com/lightningnetwork/lnd/channeldb/migration13.MigrateMPP in /home/ubuntu/lnd/channeldb/migration13/migration.go
  370.24MB     1.52GB (flat, cum) 99.75% of Total
         .          .     58:	if paymentsBucket == nil {
         .          .     59:		return nil
         .          .     60:	}
         .          .     61:
         .          .     62:	var paymentKeys [][]byte
         .    13.55MB     63:	err := paymentsBucket.ForEach(func(k, v []byte) error {
         .          .     64:		paymentKeys = append(paymentKeys, k)
         .          .     65:		return nil
         .          .     66:	})
         .          .     67:	if err != nil {
         .          .     68:		return err
         .          .     69:	}
         .          .     70:
         .          .     71:	// With all keys retrieved, start the migration.
         .          .     72:	for _, k := range paymentKeys {
         .   159.44MB     73:		bucket := paymentsBucket.NestedReadWriteBucket(k)
         .          .     74:
         .          .     75:		// We only expect sub-buckets to be found in
         .          .     76:		// this top-level bucket.
         .          .     77:		if bucket == nil {
         .          .     78:			return fmt.Errorf("non bucket element in " +
         .          .     79:				"payments bucket")
         .          .     80:		}
         .          .     81:
         .          .     82:		// Fetch old format creation info.
         .     5.50MB     83:		creationInfo := bucket.Get(paymentCreationInfoKey)
         .          .     84:		if creationInfo == nil {
         .          .     85:			return fmt.Errorf("creation info not found")
         .          .     86:		}
         .          .     87:
         .          .     88:		// Make a copy because bbolt doesn't allow this value to be
         .          .     89:		// changed in-place.
   28.50MB    28.50MB     90:		newCreationInfo := make([]byte, len(creationInfo))
         .          .     91:		copy(newCreationInfo, creationInfo)
         .          .     92:
         .          .     93:		// Convert to nano seconds.
         .          .     94:		timeBytes := newCreationInfo[32+8 : 32+8+8]
         .          .     95:		time := byteOrder.Uint64(timeBytes)
         .          .     96:		timeNs := time * 1000000000
         .          .     97:		byteOrder.PutUint64(timeBytes, timeNs)
         .          .     98:
         .          .     99:		// Write back new format creation info.
         .   227.04MB    100:		err := bucket.Put(paymentCreationInfoKey, newCreationInfo)
         .          .    101:		if err != nil {
         .          .    102:			return err
         .          .    103:		}
         .          .    104:
         .          .    105:		// No migration needed if there is no attempt stored.
         .        5MB    106:		attemptInfo := bucket.Get(paymentAttemptInfoKey)
         .          .    107:		if attemptInfo == nil {
         .          .    108:			continue
         .          .    109:		}
         .          .    110:
         .          .    111:		// Delete attempt info on the payment level.
         .        3MB    112:		if err := bucket.Delete(paymentAttemptInfoKey); err != nil {
         .          .    113:			return err
         .          .    114:		}
         .          .    115:
         .          .    116:		// Save attempt id for later use.
         .          .    117:		attemptID := attemptInfo[:8]
         .          .    118:
         .          .    119:		// Discard attempt id. It will become a bucket key in the new
         .          .    120:		// structure.
         .          .    121:		attemptInfo = attemptInfo[8:]
         .          .    122:
         .          .    123:		// Append unknown (zero) attempt time.
         .          .    124:		var zero [8]byte
  337.24MB   337.24MB    125:		attemptInfo = append(attemptInfo, zero[:]...)
         .          .    126:
         .          .    127:		// Create bucket that contains all htlcs.
         .   220.02MB    128:		htlcsBucket, err := bucket.CreateBucket(paymentHtlcsBucket)
         .          .    129:		if err != nil {
         .          .    130:			return err
         .          .    131:		}
         .          .    132:
         .          .    133:		// Create an htlc for this attempt.
         .   347.04MB    134:		htlcBucket, err := htlcsBucket.CreateBucket(attemptID)
         .          .    135:		if err != nil {
         .          .    136:			return err
         .          .    137:		}
         .          .    138:
         .          .    139:		// Save migrated attempt info.
         .   129.01MB    140:		err = htlcBucket.Put(htlcAttemptInfoKey, attemptInfo)
         .          .    141:		if err != nil {
         .          .    142:			return err
         .          .    143:		}
         .          .    144:
         .          .    145:		// Migrate settle info.
         .     2.50MB    146:		settleInfo := bucket.Get(paymentSettleInfoKey)
         .          .    147:		if settleInfo != nil {
         .          .    148:			// Payment-level settle info can be deleted.
         .          .    149:			err := bucket.Delete(paymentSettleInfoKey)
         .          .    150:			if err != nil {
         .          .    151:				return err
         .          .    152:			}
         .          .    153:
         .          .    154:			// Append unknown (zero) settle time.
         .          .    155:			settleInfo = append(settleInfo, zero[:]...)
         .          .    156:
         .          .    157:			// Save settle info.
         .   512.06kB    158:			err = htlcBucket.Put(htlcSettleInfoKey, settleInfo)
         .          .    159:			if err != nil {
         .          .    160:				return err
         .          .    161:			}
         .          .    162:
         .          .    163:			// Migration for settled htlc completed.
         .          .    164:			continue
         .          .    165:		}
         .          .    166:
         .          .    167:		// If there is no payment-level failure reason, the payment is
         .          .    168:		// still in flight and nothing else needs to be migrated.
         .          .    169:		// Otherwise the payment-level failure reason can remain
         .          .    170:		// unchanged.
         .        4MB    171:		inFlight := bucket.Get(paymentFailInfoKey) == nil
         .          .    172:		if inFlight {
         .          .    173:			continue
         .          .    174:		}
         .          .    175:
         .          .    176:		// The htlc failed. Add htlc fail info with reason unknown. We
         .          .    177:		// don't have access to the original failure reason anymore.
    4.50MB     4.50MB    178:		failInfo := []byte{
         .          .    179:			// Fail time unknown.
         .          .    180:			0, 0, 0, 0, 0, 0, 0, 0,
         .          .    181:
         .          .    182:			// Zero length wire message.
         .          .    183:			0,
         .          .    184:
         .          .    185:			// Failure reason unknown.
         .          .    186:			0,
         .          .    187:
         .          .    188:			// Failure source index zero.
         .          .    189:			0, 0, 0, 0,
         .          .    190:		}
         .          .    191:
         .          .    192:		// Save fail info.
         .    67.51MB    193:		err = htlcBucket.Put(htlcFailInfoKey, failInfo)
         .          .    194:		if err != nil {
         .          .    195:			return err
         .          .    196:		}
         .          .    197:	}
         .          .    198:
ROUTINE ======================== github.com/lightningnetwork/lnd/channeldb/migration13.MigrateMPP.func1 in /home/ubuntu/lnd/channeldb/migration13/migration.go
   13.55MB    13.55MB (flat, cum)  0.87% of Total
         .          .     59:		return nil
         .          .     60:	}
         .          .     61:
         .          .     62:	var paymentKeys [][]byte
         .          .     63:	err := paymentsBucket.ForEach(func(k, v []byte) error {
   13.55MB    13.55MB     64:		paymentKeys = append(paymentKeys, k)
         .          .     65:		return nil
         .          .     66:	})
         .          .     67:	if err != nil {
         .          .     68:		return err
         .          .     69:	}

```

File: channeldb/migration13/migration.go [hunk 3]
```diff
@@ -114,15 +134,12 @@ func MigrateMPP(tx kvdb.RwTx) error {
...
-		// structure.
-		attemptInfo = attemptInfo[8:]
+		copy(attemptID[:], attemptInfo[:8])
 
-		// Append unknown (zero) attempt time.
-		var zero [8]byte
-		attemptInfo = append(attemptInfo, zero[:]...)
+		// Copy over the attempt info into a new slice, the last 8
+		// bytes will be zero to mark a zero attempt time.
+		newAttemptInfo := make([]byte, len(attemptInfo))
```
Comment (Roasbeef) [inline channeldb/migration13/migration.go:141]: I think we can use a buffer pool both here and above, assuming the `Put` method copies the bytes over.

Comment (Crypt-iQ) [reply channeldb/migration13/migration.go:141]: `Put` method doesn't copy the `value` bytes over, it does copy the keys over: https://github.com/etcd-io/bbolt/blob/master/node.go#L139

**Comment by joostjager** [comment]
Can the migration problem be resolved by not using transactions? Make a (disk) copy of the database, apply all migrations without (big) transactions and if no error, delete old db and rename new?

**Comment by Crypt-iQ** [comment]
> Can the migration problem be resolved by not using transactions? Make a (disk) copy of the database, apply all migrations without (big) transactions and if no error, delete old db and rename new?

This would work, could also split up transactions into smaller, less resource intensive ones. If the migration is to a new bucket, and the migration fails, the next run could just delete the new bucket and try again.

**Comment by Roasbeef** [comment]
> Can the migration problem be resolved by not using transactions? Make a (disk) copy of the database, apply all migrations without (big) transactions and if no error, delete old db and rename new?

What does this give us that they dry run mode doesn't? In the end, we need to commit it all in a single transaction.

**Comment by joostjager** [comment]
That you don't need to run it all in one transaction. Copying over the migrated database _is_ the transaction.

**Comment by Roasbeef** [comment]
Gotcha makes sense, for this I think we just want to get down the memory usage to something that works with older legacy nodes that have numerous payments, if possible. Otherwise we can consider that route, but then we create an entirely new project. Also sounds like this would possibly better be implemented as a standalone tool? Also possibly related to #4131, as we might as well compact the database if we're going to copy it.

**Comment by joostjager** [comment]
I wouldn't call it an entirely new project. The copy can be made before executing any of the migrations. That way the older migrations don't need any changes. Then we can create a new migration interface without a transaction and put that on migration 13. Finally copy over the database, indeed after possibly compacting. Or better, do some renames to make the replace operation atomic.

**Comment by Roasbeef** [comment]
By new project I just mean that we can possibly look into doing it for _new_ migrations going forward. All we're after here is memory optimizations for the existing migrations. Also as it would be an entirely new way of doing migrations, it would need to have an adequate testing/vetting phase.

**Pull request closed without merging**